### DISCUSSION TITLE
About behavior HTTPException of sub-applications

### DISCUSSION THREAD

**Discussion by yoshiya0503** [fastapi/fastapi#6291] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Depends, HTTPException
from fastapi.responses import JSONResponse
from starlette.middleware.cors import CORSMiddleware


app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=['*'],
    allow_credentials=True,
    allow_methods=['*'],
    allow_headers=['*'],
    expose_headers=['*'],
)

sub = FastAPI()


@sub.get('/', status_code=200)
async def get():
    raise HTTPException(400, 'error handling sub app')


async def http_exception_handler(_, exc):
    return JSONResponse(exc.detail, status_code=exc.status_code)

# not working
app.add_exception_handler(HTTPException, http_exception_handler)

# working
# sub.add_exception_handler(HTTPException, http_exception_handler)

app.mount('/sub', sub)


@app.get('/healthcheck', status_code=200)
async def healthcheck():
    return 'health check'
```

### Description

I expect response of GET /sub endpoint is "error handling sub app".

But I got {detail: "error handling sub app"}.

The cause of this, http exception handler attached to root app will not handling sub-app exceptions.

Actually, http exception handler attached to sub app worked fine.

Is this intended behavior?

### Operating System

macOS

### Operating System Details

MacOS Monterey v 12.3.1

### FastAPI Version

0.80.0

### Python Version

Python 3.10.4

**Comment by iudeen** [comment]
I think it should be that way. Your exception handlers should be at app level, and not global?🤔

**Comment by Kludex** [comment, marked as answer]
It's intended.

**Comment by yoshiya0503** [comment]
fm...

If the middleware attached to root app, effects are reach sub app, But custom error handling will not reach.

I think error handling behavior should be same range as middleware, what do you think?

(or specify error handling range in here. https://fastapi.tiangolo.com/advanced/sub-applications/)

**Comment by iudeen** [comment]
> I think error handling behavior should be same range as middleware, what do you think?

I'm not sure if this is the right justification. I think middleware sits between your apps and the server, while your exception handlers are within your app's scope.

**Comment by yoshiya0503** [comment]
Sure, I'm not particularly troubled at this behavior, If that is intended, It's ok.
